### PR TITLE
Fix errors in MSP display port on ChibiOS platforms. This fixes HDZERO issues in bug 20955

### PR DESCRIPTION
**Pull request by pbreed**
Fixed multiple issues preventing MSP packets from being correctly processed. from HDZERO VTX, likely also effected other VTXes. See detailed description in bug 20955

### CODE REVIEW

File: libraries/AP_MSP/AP_MSP_Telem_Backend.cpp [hunk 1]
```diff
@@ -348,11 +348,14 @@ void AP_MSP_Telem_Backend::process_incoming_data()
     if (numc > 0) {
         // Process incoming bytes
         while (numc-- > 0) {
-            const uint8_t c = _msp_port.uart->read();
-            msp_parse_received_data(&_msp_port, c);
+			//Read result is signed need to check
+            const int16_t c = _msp_port.uart->read(); 
+			if(c>=0){
```
Comment (peterbarker) [inline libraries/AP_MSP/AP_MSP_Telem_Backend.cpp:353]:
```suggestion
            const int16_t c = _msp_port.uart->read(); 
			if (c < 0) {
			    break;
			}
```

File: libraries/AP_MSP/AP_MSP_Telem_Backend.cpp [hunk 2]
```diff
@@ -348,11 +348,14 @@ void AP_MSP_Telem_Backend::process_incoming_data()
...
+					msp_parse_received_data(&_msp_port,(uint8_t)c);
 
-            if (_msp_port.c_state == MSP_COMMAND_RECEIVED) {
+             if (_msp_port.c_state == MSP_COMMAND_RECEIVED) {
```
Comment (peterbarker) [inline libraries/AP_MSP/AP_MSP_Telem_Backend.cpp:356]:
```suggestion
            if (_msp_port.c_state == MSP_COMMAND_RECEIVED) {
```

File: libraries/AP_OSD/AP_OSD_MSP_DisplayPort.cpp [hunk 3]
```diff
@@ -45,8 +45,7 @@ bool AP_OSD_MSP_DisplayPort::init(void)
         gcs().send_text(MAV_SEVERITY_WARNING,"MSP DisplayPort uart not available");
         return false;
     }
-    // re-init port here for use in this thread
-    _displayport->init_uart();
+	_bInitedUart=false;
```
Comment (peterbarker) [inline libraries/AP_OSD/AP_OSD_MSP_DisplayPort.cpp:48]:
```suggestion
	  _bInitedUart=false;
```

File: libraries/AP_OSD/AP_OSD_MSP_DisplayPort.cpp [hunk 4]
```diff
@@ -77,6 +76,13 @@ uint8_t AP_OSD_MSP_DisplayPort::format_string_for_osd(char* buff, uint8_t size,
 
 void AP_OSD_MSP_DisplayPort::flush(void)
 {
+    if(!_bInitedUart)
+	{
```
Comment (peterbarker) [inline libraries/AP_OSD/AP_OSD_MSP_DisplayPort.cpp:80]:
```suggestion
    if (!_bInitedUart) {
```

... and adjust indenting

File: libraries/AP_OSD/AP_OSD_MSP_DisplayPort.h [hunk 5]
```diff
@@ -204,7 +204,7 @@ class AP_OSD_MSP_DisplayPort : public AP_OSD_Backend
         SYM_RNGFD,
         SYM_LQ,
     };
-
+	bool _bInitedUart;
```
Comment (peterbarker) [inline libraries/AP_OSD/AP_OSD_MSP_DisplayPort.h:207]:
```suggestion
    bool _bInitedUart;
```

**Review by peterbarker — CHANGES_REQUESTED** [review]
I can't comment on the protocol fixes themselves.

File: libraries/AP_MSP/AP_MSP_Telem_Backend.cpp [hunk 6]
```diff
@@ -348,11 +348,16 @@ void AP_MSP_Telem_Backend::process_incoming_data()
     if (numc > 0) {
         // Process incoming bytes
         while (numc-- > 0) {
-            const uint8_t c = _msp_port.uart->read();
-            msp_parse_received_data(&_msp_port, c);
+			//Read result is signed need to check
+            const int16_t c = _msp_port.uart->read(); 
+			if (c < 0) {
```
Comment (andyp1per) [inline libraries/AP_MSP/AP_MSP_Telem_Backend.cpp:353]: Did you try without this check? available is checked above and the call is in the same thread as this one

Comment (andyp1per) [reply libraries/AP_MSP/AP_MSP_Telem_Backend.cpp:353]: Need to format correctly

File: libraries/AP_MSP/AP_MSP_Telem_Backend.cpp [hunk 7]
```diff
@@ -348,11 +348,16 @@ void AP_MSP_Telem_Backend::process_incoming_data()
...
+			    break;
+			}
+					msp_parse_received_data(&_msp_port,(uint8_t)c);
 
             if (_msp_port.c_state == MSP_COMMAND_RECEIVED) {
                 msp_process_received_command();
+			 }
```
Comment (andyp1per) [inline libraries/AP_MSP/AP_MSP_Telem_Backend.cpp:360]: formatting

File: libraries/AP_MSP/msp.cpp [hunk 8]
```diff
@@ -177,7 +177,7 @@ bool MSP::msp_parse_received_data(msp_port_t *msp, uint8_t c)
         break;
 
     case MSP_HEADER_M:      // Waiting for '<'
-        if (c == '<') {
+        if ((c == '<')|| (c=='>')) {
```
Comment (andyp1per) [inline libraries/AP_MSP/msp.cpp:180]: This indicates a packet in the opposite direction and appears wrong to me. If correct the comment needs fixing as well

File: libraries/AP_OSD/AP_OSD_MSP_DisplayPort.cpp [hunk 9]
```diff
@@ -77,6 +76,12 @@ uint8_t AP_OSD_MSP_DisplayPort::format_string_for_osd(char* buff, uint8_t size,
 
 void AP_OSD_MSP_DisplayPort::flush(void)
 {
+    if (!_bInitedUart) {
```
Comment (andyp1per) [inline libraries/AP_OSD/AP_OSD_MSP_DisplayPort.cpp:79]: Can you explain this? There are some threading issues in the way the uart is set up, but I am not sure which one you are trying to fix here.

Needs formatting

**Review by andyp1per — CHANGES_REQUESTED** [review]

File: libraries/AP_OSD/AP_OSD_MSP_DisplayPort.cpp (hunk 9 above)
Comment (pbreed) [reply libraries/AP_OSD/AP_OSD_MSP_DisplayPort.cpp:79]: If the internal uart variable thread owner does not match the current thread read function fails.
Look at the read code. TSo uart init must get called from the thread of the task that is going to do the actual reading.
As it was written before uart init was called from the scope of a differnt thread leading to the receive not working.
he available does not have a thread check since its a const function,
but read does.  This leads to an infinite loop where availible says you have data, but read returns -1 as the read thread check fails. The fact this could get messed up in the future is why I added the -1 check to the read function in the other file.
As for format do you guys have clang formatter  template? I'm an allman  brace code guy and I find the traiilng brace really hard to read... and have no idea what needs done to fix formatting to your standard neither the brace placement nor indenting make any sense to me at all it looks like random gibberish to me... my ability to fix formatting to your standard is basically non existent,

File: libraries/AP_MSP/msp.cpp (hunk 8 above)
Comment (pbreed) [reply libraries/AP_MSP/msp.cpp:180]: On the HDZero the direction was opposite, and the parser failed.

File: libraries/AP_MSP/AP_MSP_Telem_Backend.cpp (hunk 6 above)
Comment (pbreed) [reply libraries/AP_MSP/AP_MSP_Telem_Backend.cpp:353]: If everything is working correctly the check is not needed. If the uart thread init gets messed up elsewhere the call is needed. Otherwise you get in an infite loop where available says you have data, but you can't actually read it.

Comment (pbreed) [reply libraries/AP_MSP/AP_MSP_Telem_Backend.cpp:355]: This is currently broken and not compiling... you made a logic change suggestion and I accepted the change.
The change had a brace matching error...
I've fixed it a pushed to my personal branch... not sure what the process is to get the fix back over here to this pull?

The whole github pull thing is new to me as I have direct write privileges' in the git codebase I work on for a living.
This actually the first pull request I've ever done.....

Comment (andyp1per) [reply libraries/AP_MSP/AP_MSP_Telem_Backend.cpp]: https://ardupilot.org/dev/docs/submitting-patches-back-to-master.html

File: libraries/AP_OSD/AP_OSD_MSP_DisplayPort.cpp (hunk 9 above)
Comment (andyp1per) [reply libraries/AP_OSD/AP_OSD_MSP_DisplayPort.cpp:79]: Ok, makes sense but needs a comment

https://ardupilot.org/dev/docs/style-guide.html#style-guide

File: libraries/AP_MSP/AP_MSP_Telem_Backend.cpp [hunk 10]
```diff
@@ -584,9 +591,9 @@ MSPCommandResult AP_MSP_Telem_Backend::msp_process_out_api_version(sbuf_t *dst)
         uint8_t major;
         uint8_t minor;
     } api_version  {
-        proto : MSP_PROTOCOL_VERSION,
```
Comment (andyp1per) [inline libraries/AP_MSP/AP_MSP_Telem_Backend.cpp:587]: Yeah, obviously gratuitous whitespace changes are a no-no :)

File: libraries/AP_MSP/msp.cpp (hunk 8 above)
Comment (andyp1per) [reply libraries/AP_MSP/msp.cpp:180]: It makes some sense, but also seems counter to what the function is trying to do and it's name - might be worth checking betaflight/src/main/msp/msp_serial.c from which this was ported to see if that code has evolved in a compatible way

File: libraries/AP_OSD/AP_OSD_MSP_DisplayPort.cpp (hunk 9 above)
Comment (andyp1per) [reply libraries/AP_OSD/AP_OSD_MSP_DisplayPort.cpp:79]: There is no problem with available() being const because the thread ownership function is in C (actually it might just be a C) macro. That said it is *not* const - so this argument doesn't apply. Either way they *must* be consistent so https://github.com/ArduPilot/ardupilot/pull/20963 is the right way to do this

File: libraries/AP_MSP/AP_MSP_Telem_Backend.cpp (hunk 10 above)
Comment (pbreed) [reply libraries/AP_MSP/AP_MSP_Telem_Backend.cpp:587]: Trying to fix formatting I ran the astyle tool with the style sheet checked in under tools and submitted that...
So other than that not sure how to resolve the formatting issues...

**Review by pbreed — COMMENTED** [review]
Its all kinda hard to follow.
I think I've fixed everything there is to be fixed.
For formatting I ran astyle with the checked in formatter....
So gratuitus white space changed seem to be from that...

File: libraries/AP_MSP/AP_MSP_Telem_Backend.cpp (hunk 6 above)
Comment (peterbarker) [reply libraries/AP_MSP/AP_MSP_Telem_Backend.cpp]: > If everything is working correctly the check is not needed. If the uart thread init gets messed up elsewhere the call is needed. Otherwise you get in an infite loop where available says you have data, but you can't actually read it.

Mmm.... I can't actually see this infinite loop.  `numc` is only assigned to once, and decremented regardless of `read()`'s return value.  Sure, we're probably going to get the same problem next time this method's called, but I don't *think* this is a fall-out-of-sky bug.  Assigning into a `numc` like this is an idiom in our code to avoid the sorts of loop you're thinking of...

I like this top patch - would you mind if I extracted it into a separate PR? (maintaining authorship, of course).

Comment (pbreed) [reply libraries/AP_MSP/AP_MSP_Telem_Backend.cpp]: Eros saves from crash, but from thread task view it’s an infinite loop running 1024 every 10 msec to no positive effect. I’m away from my computer for next 7 days. So feel free to use any part of this. With your available changes only msp.cop and AP_OSD_MSP_Backend changes needed to fix hdzero ops. Don’t care about anything other than having it functional. Thanks for taking time to review this.

**Comment by rmackay9** [comment]
@pbreed, txs for the contribution.

By the way, we normally prefix all our commits with the name of the library or vehicle.  So it would be good if you could prefix them to start with "AP_MSP:" or "AP_OSD:".

**Comment by pbreed** [comment]
Not infinite in that it's limited to 1024 times per 100hz, so rtos is
saving crash, but burning cycles to no positive effect...infinite loop from
this tasks/thread perspective.

I'm away from my computer for next 7vdays, so do anything you want to do
with any part of this feel free...just want it fixed...
